MSC3970: Scope transaction IDs to devices (#3970)
matrix-org/matrix-spec-proposals: proposals/3970-device-scope-txnid.md

The Synapse homeserver only supports local echo (by the presence of `transaction_id` on sync responses) for the previous
24 hours. This is because the `event_txn_id` table is only kept for 24 hours.

Again, this suggests that in reality the local echo semantics are not preserved eternally as the spec might suggest.

### Is the "device" concept the right level of abstraction to use?

One way to look at it is that device is already widely used in the end-to-end encryption parts of the spec and so why
isn't it suitable for this use case too?

### What about two clients masquerading as a single device ID?

I don't know if this actually works in practice. If this was a concern then it could be mitigated by clarifying in the
spec that if a client wishes to submit requests using the same `device_id` as another client session that it should
choose transaction identifiers that are unique to that client session.

## Alternatives

### Do nothing

We could leave the transaction ID scope as is.

However, it makes it difficult to implement a features like [MSC3861: Matrix architecture change to delegate authentication via OIDC](https://github.com/matrix-org/matrix-spec-proposals/pull/3861)

> as the concept of a "client session" doesn't really exist in OIDC.

Author: There are two parts to this in the [OIDC proposal](https://github.com/matrix-org/matrix-spec-proposals/pull/3861):

1. The client generates their own device ID and specifies it in the OIDC scope that is requested at the point of sign in. This scope described in [MSC2967](https://github.com/sandhose/matrix-doc/blob/msc/sandhose/api-scopes/proposals/2967-api-scopes.md#device-id-handling)
2. When the client signs out the OpenID Provider is responsible for informing the Homeserver that the device should be deleted. This process is described in [MSC3964](https://github.com/sandhose/matrix-doc/blob/msc/sandhose/oauth2-profile/proposals/2964-oauth2-profile.md#logging-out)

- [`PUT /_matrix/client/v3/rooms/{roomId}/redact/{eventId}/{txnId}`](https://spec.matrix.org/v1.6/client-server-api/#put_matrixclientv3roomsroomidredacteventidtxnid)
- [`PUT /_matrix/client/v3/sendToDevice/{eventType}/{txnId}`](https://spec.matrix.org/v1.6/client-server-api/#put_matrixclientv3sendtodeviceeventtypetxnid)

## Proposal

It is proposed that the scope of transaction identifiers be changed from a
"client session" to a "device".

A "device" is typically represented by a `device_id` elsewhere in the spec.

For idempotency, this means the homeserver changing the method of identifying a
request from:

- (`client session`, `HTTP path of request which includes the transaction ID`)

to:

- (`device_id`, `HTTP path of request which includes the transaction ID`)

For local echo, the homeserver would now include the `transaction_id` in the
event data when it is serving a sync request from the same `device_id` as
determined from the access token.

## Potential issues

### This is technically a breaking change to the spec

The main "issue" I see with this proposal is that this is technically a breaking
change to the spec.

> Because a device ID could have multiple sequences of access tokens associated
> with it, this proposal would widen the scope of the transaction ID.

Reviewer: I'm somewhat struggling to follow this explanation. I think an example of how it might happen in practice is illustrative:

```suggestion
A device ID could have multiple sequences of access tokens associated
with it (since device ID can be specified as a parameter to 
[`/login`](https://spec.matrix.org/v1.6/client-server-api/#post_matrixclientv3login)). 

For example, a "bot" implementation might masquerade as the same "device" despite
calling `/login` on every run by passing the same device ID. Such a device might
also use the same transaction ID on each run.
```

> Therefore it could potentially lead to a request being treated as "new" where
> before it would have been identified as a retransmission and deduplicated.

Reviewer: the opposite, no?

```suggestion
Therefore it could potentially lead to a request being treated as a duplicate
where previously it would have been treated as a distinct request.
```

Author: You are quite right 👍

> However, the evidence suggests that nothing would be impacted in reality.
> 
> Some evidences have been collated to support this claim:

Reviewer: ```suggestion
However, some evidence has been collated which suggests that nothing would be impacted
in reality:
```

> #### 1. Data from the matrix.org Homeserver suggests the change would have no impact

Reviewer: ```suggestion
#### 1. Data from the matrix.org homeserver suggests the change would have no impact
```

> The matrix.org Homeserver is a reasonable size deployment and could be considered
> reasonably representative of the diversity of Matrix client.

Reviewer: ```suggestion
The `matrix.org` homeserver is a reasonable size deployment and could be considered
reasonably representative of the diversity of Matrix clients.
```

> The Synapse homeserver that runs matrix.org maintains a `event_txn_id` table
> that contains a rolling 24 hour window of
> (`user_id`, `token_id`, `room_id`, `txn_id`) tuples.

Reviewer: ```suggestion
Synapse maintains a `event_txn_id` table
that contains a rolling 24 hour window of
(`user_id`, `token_id`, `room_id`, `txn_id`) tuples.
```

Having analysed the contents of the table, it appears that there are no repeated
transaction IDs for a given user, token and room.

n.b. not all `PUT` endpoints contribute to the table but I think the high-volume
ones do

> As such, the widening of the scope from token to device should not have caused
> any issues during the periods sampled.

Reviewer: ```suggestion
This suggests that the widening of the scope from token to device would not have caused
any issues during the periods sampled.
```

- [`PUT /_matrix/client/v3/rooms/{roomId}/send/{eventType}/{txnId}`](https://spec.matrix.org/v1.6/client-server-api/#put_matrixclientv3roomsroomidsendeventtypetxnid)
- [`PUT /_matrix/client/v3/rooms/{roomId}/redact/{eventId}/{txnId}`](https://spec.matrix.org/v1.6/client-server-api/#put_matrixclientv3roomsroomidredacteventidtxnid)
- [`PUT /_matrix/client/v3/sendToDevice/{eventType}/{txnId}`](https://spec.matrix.org/v1.6/client-server-api/#put_matrixclientv3sendtodeviceeventtypetxnid)

## Proposal

It is proposed that the scope of transaction identifiers be changed from a
"client session" to a "device".

A "device" is typically represented by a `device_id` elsewhere in the spec.

For idempotency, this means the homeserver changing the method of identifying a
request from:

- (`client session`, `HTTP path of request which includes the transaction ID`)

to:

- (`device_id`, `HTTP path of request which includes the transaction ID`)

For local echo, the homeserver would now include the `transaction_id` in the
event data when it is serving a sync request from the same `device_id` as
determined from the access token.

## Potential issues

> ### This is technically a breaking change to the spec

Reviewer: @turt2live I think you are the most likely person on the SCT to have concerns over this approach - could you have a look?

support both semantics for a period of time for (empirically) no gain in return.

## Security considerations

A malicious client can adopt an existing device ID of a user. This could
possibly allow some kind of denial of service attack.

However, if such an attack where possible it would be possible to do so without
this MSC as device IDs are crucial to the implementation of end-to-end encryption.

## Other recommendations

I'm not suggesting that these recommendations are address in this proposal, but
more notes for future proposals or spec clarifications.

### Clarification on idempotency semantics

I have separately prepared a [spec PR](https://github.com/matrix-org/matrix-spec/pull/1449)
to clarify some of the idempotency semantics that doesn't modify the spec but is
useful to understand the context of this proposal.

### Clarification on transaction ID time scope

I also suggest that the spec be clarified over what time periods the transaction
ID is scoped for such that clients can be aware. This cloud simply be to say
that the time period is not defined and so may vary by implementation.

### Suggest a lease naive transaction ID format

> I think we should sense to update the recommendation on the format of

Author: @uhoreg you are not the only one. I attempted to re-read this morning and couldn't make sense of it either!

I've reworked it into something that hopefully reads better.

> # MSC3970: Scope transaction IDs to devices

Reviewer 2: As someone with very little context on what triggered this MSC to be proposed, I find it quite difficult to figure out the problem it's trying to solve just from reading the MSC's introduction. The "Do nothing" part of the Alternatives section has a bit to say about that, but I think it would be helpful to also have this information here, so by the time the reader gets to the proposal it's clear what problem it's trying to solve.

Author: I've had a re-read and I agree. I've added some more info in the intro section to try and make it clearer sooner: [`6fc4975` (#3970)](https://github.com/matrix-org/matrix-spec-proposals/pull/3970/commits/6fc4975c52ec3ed72b9234aa01c81551850c840d)

Reviewer 2: This is very helpful, thanks!
